server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -612,4 +634,113 @@ private void failShard(ReplicationFailedException e, IndexShard indexShard) {
+                IndexShard replicaShard = this.indicesService.getShardOrNull(shardId);
+                if (replicaShard != null && replicaShard.state() == IndexShardState.STARTED) {
+                    failShard(new ReplicationFailedException("replica too far behind primary, marking as stale"), replicaShard);
+                }
+            }
+        }
+
+        @Override
+        protected String getThreadPool() {
+            return ThreadPool.Names.GENERIC;
+        }
+
+        @Override
+        public String toString() {
+            return "fail_stale_replica";
+        }
+
+    }
+
+    /**
+     * Internal class to track the list of Replication Checkpoints which are still not processed.
+     */
+    public static class PendingCheckpoints {
+        protected final Map<ShardId, List<Tuple<ReplicationCheckpoint, Long>>> checkpointsTracker = ConcurrentCollections
+            .newConcurrentMap();
+        private volatile TimeValue maxAllowedReplicationTime;
+
+        public PendingCheckpoints(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void setMaxAllowedReplicationTime(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void addNewReceivedCheckpoint(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            checkpointsTracker.putIfAbsent(shardId, new ArrayList<>());
+            List<Tuple<ReplicationCheckpoint, Long>> checkpoints = checkpointsTracker.get(shardId);
+            if (checkpoints.size() > 0) {
+                assert checkpoints.get(checkpoints.size() - 1).v1().isAheadOf(checkpoint) == false;
+            }
+            checkpoints.add(new Tuple<>(checkpoint, System.currentTimeMillis()));
+        }
+
+        public void updateCheckpointProcessed(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            if (checkpointsTracker.containsKey(shardId)) {
+                checkpointsTracker.put(
+                    shardId,
+                    checkpointsTracker.get(shardId)
+                        .stream()
+                        .filter(t -> t.v1().isAheadOf(checkpoint))
+                        .collect(Collectors.toCollection(ArrayList::new))
+                );
+            }
+        }
+
+        public ReplicationCheckpoint getLatestReplicationCheckpoint(ShardId shardId) {
+            List<Tuple<ReplicationCheckpoint, Long>> checkpoints = checkpointsTracker.getOrDefault(shardId, Collections.emptyList());
+            if (checkpoints.size() > 0) {
+                return checkpoints.get(checkpoints.size() - 1).v1();
+            }
+            return null;
+        }
+
+        public void remove(ShardId shardId) {
+            checkpointsTracker.remove(shardId);
+        }
+
+        public List<ShardId> getStaleShardsToFail() {
+            long currentTime = System.currentTimeMillis();
```

Comment:
Lets pass the time as a param so we aren't computing it for every shard.

Reply:
This one is running once per node so we're only computing it once.

Reply:
To calculate elapsed time always use the logical clock `System.nanoTime` instead of the wall clock time `currentTimeMillis`

Reply:
Curious if we are basing staleness based on time, how do we prevent GC pauses from not adversely impacting shards hosted on the node. Can it cause multiple shards to now fail?

Reply:
Yeah, GC pauses can also impact the shards on the node but the default timeout of 10 minute should be pretty generous for it to be a concern. LMK what you think.

Reply:
I was exploring using `System.nanoTime` which will provide more accuracy but is also expensive compared to `currentTimeMillis`. Since we're not sensitive small time differences in this particular case, i've kept it as it is.

Reply:
It's not about accuracy its more about correctness. Elapsed time is correctly measured using logical clock and not physical clocks

Reply:
Thanks. I've updated the PR to use `System.nanoTime` now. 

> Will prefer computing staleness on other factors like bytes/checkpoints behind in addition to time

@mch2 Any thoughts on this?

We can explore introducing another setting for something like checkpoints behind. But we also have similar setting on primary which can also be confusing to user.  Thinking out loud here, maybe we can explore reusing the same setting(s) here for staleness criteria.

Reply:
yes this is a good callout a long GC pause could fail multiple shards here.  perhaps we restrict this to only fail 1 shard every few cycles?

Bytes behind today is a rough estimate, but we could use this.  It's a rough estimate because the size of a checkpoint is computed from a sum of the segments in the checkpoint.  It is not the diff of segments missing on the replica.  Meaning a checkpoint containing only a merge would not show the replica behind.  To be precise we need to include all StoreFileMetadata inside of each published checkpoint.  Although in this case the shard is not really "behind" as the docs are searchable. With that aside, the question is what would be a sane default value to use? I think we'd need to capture and use a multiple of a transfer rate or checkpoint size?  A fixed value feels arbitrary.

Reply:
Another thought here is that the intent is to fail shards only if they are egregiously behind so they are not serving search requests & force recovery.  Rather than using an async task to periodically check all shards, what if we only 1. determine if a shard is not making progress on an existing event with a generous timeout value and flag the shard if it fails to catch up after n timeouts we fail it and 2. Compute bytes/time behind only as a new checkpoint is received?

Reply:
>perhaps we restrict this to only fail 1 shard every few cycles and compute cycle time?

I was exploring this as well. Main issue with only failing one(or X) shards per node is setting a default value that works well in all cases. For e.g. if a node has 1000 shards we probably don't want to wait for 30(poll interval) * 1000 seconds for failing the last stale replica. I can keep it around 100 but that also seems pretty high. Maybe we can expose this as a percentage of shards to fail per cycle and start with `.1`(or 10%). 

> Bytes behind today is a rough estimate, but we could use this. It's a rough estimate because the size of a checkpoint is computed from a sum of the segments in the checkpoint

We can instead use the criteria of N checkpoints behind since we're already preserving all the pending checkpoints with this change. Bytes behind, like you mentioned, is a rough estimate and is tricky to set a "sane" default.

> Thinking on this more, if we have pressure enabled to prevent additional checkpoints in the group do we even need 2 above? We can simply consider ongoing replication time.

That was primary reason to keep it simple and only consider replication time originally. This along with checkpoints behind can work well I think. LMK.